compiler/rustc_mir_build/src/build/expr/as_rvalue.rs

```diff
@@ -196,17 +197,64 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                 // discriminant. Previously every backend had to repeat the logic for this operation. Now we
                 // create all the steps directly in MIR with operations all backends need to support anyway.
                 let (source, ty) = if let ty::Adt(adt_def, ..) = source.ty.kind() && adt_def.is_enum() {
+
                     let discr_ty = adt_def.repr().discr_type().to_ty(this.tcx);
-                    let temp = unpack!(block = this.as_temp(block, scope, source, Mutability::Not));
+                    let layout = this.tcx.layout_of(this.param_env.and(source.ty));
                     let discr = this.temp(discr_ty, source.span);
+                    let temp = unpack!(block = this.as_temp(block, scope, source, Mutability::Not));
                     this.cfg.push_assign(
                         block,
                         source_info,
                         discr,
                         Rvalue::Discriminant(temp.into()),
                     );
+                    let (op,ty) = (Operand::Move(discr), discr_ty);
+
+                    if let Abi::Scalar(scalar) = layout.unwrap().abi{
+                        if let Primitive::Int(_i,signed) = scalar.primitive(){
+                            let range = scalar.valid_range(&this.tcx);
+                            let max_size = scalar.size(&this.tcx).unsigned_int_max();
+                            if range.end >= range.start {
+                                let range_end = ConstantKind::from_bits(this.tcx, range.end, ty::ParamEnv::empty().and(discr_ty));
+                                let lit_op = this.literal_operand(expr.span, range_end);
+                                let bool_ty = this.tcx.types.bool;
+                                let is_ge = this.temp(bool_ty,expr_span);
+
+                                this.cfg.push_assign(
+                                    block,
+                                    source_info,
+                                    is_ge,
+                                    Rvalue::BinaryOp(BinOp::Ge, Box::new(((lit_op),(Operand::Copy(discr)))))
+                                );
+                                this.cfg.push(
+                                    block,
+                                    Statement {
+                                        source_info,
+                                        kind: StatementKind::Intrinsic(Box::new(NonDivergingIntrinsic::Assume(Operand::Copy(is_ge)))) },
+                                );
+                                if signed && range.end != max_size{
+                                    let range_start = ConstantKind::from_bits(this.tcx,range.end, ty::ParamEnv::empty().and(discr_ty));
```

Comment:
I don't get the `range.end` being used as `range_start` logic.

I think you can just use `range.start` and not have the above condition at all, just always encode a start `assume`

Reply:
good catch :sweat_smile: